src/liballoc/collections/btree/navigate.rs

```diff
@@ -3,118 +3,63 @@ use core::ptr;
 use super::unwrap_unchecked;
 
-macro_rules! def_next {
-    { unsafe fn $name:ident : $next_kv:ident $next_edge:ident $initial_leaf_edge:ident } => {
-        /// Given a leaf edge handle into an immutable tree, returns a handle to the next
-        /// leaf edge and references to the key and value between these edges.
-        /// Unsafe because the caller must ensure that the given leaf edge has a successor.
-        unsafe fn $name <'a, K: 'a, V: 'a>(
-            leaf_edge: Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Edge>,
-        ) -> (Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Edge>, &'a K, &'a V) {
-            let mut cur_handle = match leaf_edge.$next_kv() {
-                Ok(leaf_kv) => {
-                    let (k, v) = leaf_kv.into_kv();
-                    let next_leaf_edge = leaf_kv.$next_edge();
-                    return (next_leaf_edge, k, v);
-                }
-                Err(last_edge) => {
-                    let next_level = last_edge.into_node().ascend().ok();
-                    unwrap_unchecked(next_level)
+macro_rules! def_next_kv {
+    { pub fn $name:ident : $adjacent_kv:ident } => {
+        /// Given a leaf edge handle, returns [`Result::Ok`] with a handle to the KV next to it,
+        /// which is either in the leaf node or in an ancestor node.
+        /// If the leaf edge is at the end of the tree, returns [`Result::Err`] with the root node.
+        pub fn $name (
+            self,
+        ) -> Result<Handle<NodeRef<BorrowType, K, V, marker::LeafOrInternal>, marker::KV>,
+                    NodeRef<BorrowType, K, V, marker::LeafOrInternal>> {
+            let mut parent_edge = match self.$adjacent_kv() {
+                Ok(leaf_kv) => return Ok(leaf_kv.forget_node_type()),
+                Err(last_edge) => match last_edge.into_node().ascend() {
+                    Ok(parent_edge) => parent_edge,
+                    Err(root) => return Err(root.forget_type()),
                 }
             };
 
             loop {
-                cur_handle = match cur_handle.$next_kv() {
-                    Ok(internal_kv) => {
-                        let (k, v) = internal_kv.into_kv();
-                        let next_internal_edge = internal_kv.$next_edge();
-                        let next_leaf_edge = next_internal_edge.descend().$initial_leaf_edge();
-                        return (next_leaf_edge, k, v);
-                    }
-                    Err(last_edge) => {
-                        let next_level = last_edge.into_node().ascend().ok();
-                        unwrap_unchecked(next_level)
+                parent_edge = match parent_edge.$adjacent_kv() {
+                    Ok(internal_kv) => return Ok(internal_kv.forget_node_type()),
+                    Err(last_edge) => match last_edge.into_node().ascend() {
+                        Ok(parent_edge) => parent_edge,
+                        Err(root) => return Err(root.forget_type()),
```

Comment:
Could we avoid duplicating this code? It looks like we should be able to, by starting out with `let mut parent_edge = self;`

Reply:
Ok, indeed. Not quite starting from there but from:

     let mut edge = self.forget_node_type();

which has to be defined for edges too, but then it works. Benchmarking.

Reply:
Seriously? Another factor 2 boost? It's getting ridiculous.